src/lib.rs

```diff
@@ -1,40 +1,44 @@
 #![recursion_limit = "1024"]
 
-extern crate heck;
-extern crate proc_macro;
-#[macro_use]
-extern crate quote;
-extern crate syn;
 
 use heck::SnakeCase;
 use proc_macro::TokenStream;
-use quote::Tokens;
+use quote::quote;
 use syn::*;
+use proc_macro2::{Ident, Span};
 
 #[proc_macro_derive(DbEnum, attributes(PgType, DieselType, db_rename))]
 pub fn derive(input: TokenStream) -> TokenStream {
-    let input = input.to_string();
-    let ast = syn::parse_derive_input(&input).expect("Failed to parse item");
-    let db_type =
-        type_from_attrs(&ast.attrs, "PgType").unwrap_or(ast.ident.as_ref().to_snake_case());
-    let diesel_mapping = type_from_attrs(&ast.attrs, "DieselType")
-        .unwrap_or(format!("{}Mapping", ast.ident.as_ref()));
-    let diesel_mapping = Ident::new(diesel_mapping);
-
-    let quoted = if let Body::Enum(ref variants) = ast.body {
-        generate_derive_enum_impls(&db_type, &diesel_mapping, &ast.ident, variants)
+    let input:DeriveInput =  parse_macro_input!(input as DeriveInput);
+    let db_type = 
+        type_from_attrs(&input.attrs, "PgType")
+            .unwrap_or(input.ident.to_string().to_snake_case());
+    let diesel_mapping = 
+        type_from_attrs(&input.attrs, "DieselType")
+            .unwrap_or(format!("{}Mapping", input.ident));
+
+    let diesel_mapping = Ident::new(diesel_mapping.as_ref(), Span::call_site());
+    let quoted = if let Data::Enum(syn::DataEnum{variants: data_variants, ..})= input.data {
+        generate_derive_enum_impls(&db_type, &diesel_mapping, &input.ident, &data_variants)
     } else {
         panic!("#derive(DbEnum) can only be applied to enums")
```

Comment:
Not really on topic, but since you've started caring about spans and removed panics from above, might as well replace this with something like `return syn::Error::new(input.span(), "derive(DbEnum) can only be applied to enums).to_compile_error()`

Reply:
Thank you, like I stated this was my first time working in any sort of macros. I really want better errors to be standard in crates that offer macros.

Reply:
I have also only recently discovered how to properly report errors, the `syn` crate has some documentation on it: https://docs.rs/syn/1.0.17/syn/struct.Error.html

By the way your code looks better than my first attempt at proc macros, so don't understate your coding skills!